baseapp/baseapp_test.go

```diff
@@ -98,6 +97,56 @@ func setupBaseApp(t *testing.T, options ...AppOption) *BaseApp {
 	return app
 }
 
+// baseapp loaded from a fraudproof
+func setupBaseAppFromFraudProof(t *testing.T, fraudProof FraudProof, options ...AppOption) *BaseApp {
+	storeKeys := make([]stypes.StoreKey, 0, len(fraudProof.stateWitness))
+	routerOpts := make([]func(bapp *BaseApp), 0, len(fraudProof.stateWitness))
+	for storeKeyName := range fraudProof.stateWitness {
+		storeKey := sdk.NewKVStoreKey(storeKeyName)
+		storeKeys = append(storeKeys, storeKey)
+
+		routerOpt := func(bapp *BaseApp) {
+			bapp.Router().AddRoute(sdk.NewRoute(routeMsgKeyValue, func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+				kv := msg.(*msgKeyValue)
+				bapp.cms.GetKVStore(storeKey).Set(kv.Key, kv.Value)
+				return &sdk.Result{}, nil
+			}))
+		}
+		routerOpts = append(routerOpts, routerOpt)
+	}
+	options = append(options, SetSubstores(storeKeys...))
+	// RouterOpts should only be called after call to `SetSubstores`
+	for _, routerOpt := range routerOpts {
+		options = append(options, AppOptionFunc(routerOpt))
+	}
+
+	// This initial height is used in `BeginBlock` in `validateHeight`
+	options = append(options, SetInitialHeight(fraudProof.blockHeight))
+
+	for storeKey := range fraudProof.stateWitness {
+		stateWitness := fraudProof.stateWitness[storeKey]
+		witnessData := stateWitness.WitnessData
+		for _, witness := range witnessData {
+			// Optimization
+			// TODO:
+			// Verify proof inside WitnessData: Not sure since canot do it before setting up without doing the redundant work on creating deepSubTrees here (ideally optimint does it)
+
+			options = append(options, SetSubstoreKVPair(storeKey, witness.Key, witness.Value))
+		}
+		// Q. How do we use stateWitness.root to ensure the root of the underlying subtree just initialized is the same? Do we need that?
```

Comment:
Is this comment/question still relevant?

Reply:
thanks for pointing out! removed it here